larpandoracontent/LArReclustering/SimplePCAThreeDClusteringTool.cc

```diff
@@ -0,0 +1,125 @@
+        (axisDirection.GetY() < std::min(axisDirection.GetX(), axisDirection.GetZ())) ? CartesianVector(0.f, 1.f, 0.f) : CartesianVector(0.f, 0.f, 1.f));
+    const CartesianVector orthoDirection1(seedDirection.GetCrossProduct(axisDirection).GetUnitVector());
+
+    //Lists for hits that have a positive and negative projection on the secondary axis
+    CaloHitList *posCaloHitList = new CaloHitList;
+    CaloHitList *negCaloHitList = new CaloHitList;
+
+    for (const CaloHit *const pCaloHit : *pInitialCaloHitList)
+    {
+        const CartesianVector pCaloHitPosition = pCaloHit->GetPositionVector();
+
+        if((pCaloHitPosition-centroid).GetDotProduct(centroid+orthoDirection1)<0)
+		{
+            negCaloHitList->push_back(pCaloHit);
+        }
+		else
+		{ 
+            posCaloHitList->push_back(pCaloHit);
+		}
+    }
+
+    newCaloHitListsVector.push_back(posCaloHitList);
+    newCaloHitListsVector.push_back(negCaloHitList);
+
+    //Now, run PCA independently on pos and neg hit lists, to refine split
+
+    //Get pos list PCA
+    CartesianVector centroidPos(0.f, 0.f, 0.f);
+    LArPcaHelper::EigenVectors eigenVecsPos;
+    LArPcaHelper::EigenValues eigenValuesPos(0.f, 0.f, 0.f);
+    LArPcaHelper::RunPca(*posCaloHitList, centroidPos, eigenValuesPos, eigenVecsPos);
+
+    // By convention, the primary axis has a positive z-component.
+    const CartesianVector axisDirectionPos(eigenVecsPos.at(0).GetZ() > 0.f ? eigenVecsPos.at(0) : eigenVecsPos.at(0) * -1.f);
+
+    //Get neg list PCA
+    CartesianVector centroidNeg(0.f, 0.f, 0.f);
+    LArPcaHelper::EigenVectors eigenVecsNeg;
+    LArPcaHelper::EigenValues eigenValuesNeg(0.f, 0.f, 0.f);
+    LArPcaHelper::RunPca(*negCaloHitList, centroidNeg, eigenValuesNeg, eigenVecsNeg);
+
+    // By convention, the primary axis has a positive z-component.
+    const CartesianVector axisDirectionNeg(eigenVecsNeg.at(0).GetZ() > 0.f ? eigenVecsNeg.at(0) : eigenVecsNeg.at(0) * -1.f);
+
+    //Get intersection point of the two new principal axes
+    CartesianVector intersectionPoint(0.f, 0.f, 0.f);
+    float displacementPos(0.f),displacementNeg(0.f);
+    LArPointingClusterHelper::GetIntersection(centroidPos,axisDirectionPos,centroidNeg,axisDirectionNeg,intersectionPoint,displacementPos,displacementNeg);
+
+    //Clear pos and neg lists
+    posCaloHitList->clear();
+    negCaloHitList->clear();
+
+    //Loop over original hit list, check whether it is within smaller cone of pos or neg axis, attach to relevant list
+    float cosConeAxisPos(0), cosConeAxisNeg(0);
+    for (const CaloHit *const pCaloHit3D : *pInitialCaloHitList)
+    {
+        cosConeAxisPos = axisDirectionPos.GetCosOpeningAngle(pCaloHit3D->GetPositionVector()-intersectionPoint);
+        cosConeAxisNeg = axisDirectionNeg.GetCosOpeningAngle(pCaloHit3D->GetPositionVector()-intersectionPoint);
+        if(cosConeAxisPos>cosConeAxisNeg) posCaloHitList->push_back(pCaloHit3D);
+        else negCaloHitList->push_back(pCaloHit3D);
+    }
```

Comment:
I don't think `float cosConeAxisPos(0), cosConeAxisNeg(0);` need to be defined outside the for loop, since they only need to exist within the scope of each iteration? They can be defined as `const float` inside the loop? Also indentation for the `if ... else` statement